server/src/main/java/org/opensearch/search/sandboxing/tracker/SandboxResourceUsageTrackerService.java

```diff
@@ -0,0 +1,123 @@
+     * @param taskResourceTrackingService Service that helps track resource usage of tasks running on a node.
+     */
+    @Inject
+    public SandboxResourceUsageTrackerService(
+        final TaskManager taskManager,
+        final TaskResourceTrackingService taskResourceTrackingService
+    ) {
+        this.taskManager = taskManager;
+        this.taskResourceTrackingService = taskResourceTrackingService;
+    }
+
+    /**
+     * Constructs a map of SandboxLevelResourceUsageView instances for each sandbox.
+     *
+     * @return Map of sandbox views
+     */
+    @Override
+    public Map<String, SandboxLevelResourceUsageView> constructSandboxLevelUsageViews() {
+        Map<String, SandboxLevelResourceUsageView> sandboxViews = new HashMap<>();
+
+        Map<String, List<Task>> tasksBySandbox = getTasksGroupedBySandbox();
+        Map<String, Map<SandboxResourceType, Long>> sandboxResourceUsage = getResourceUsageOfSandboxes(tasksBySandbox);
+
+        for (String sandboxId : tasksBySandbox.keySet()) {
+            SandboxLevelResourceUsageView sandboxLevelResourceUsageView = new SandboxLevelResourceUsageView(
+                sandboxId,
+                sandboxResourceUsage.get(sandboxId),
+                tasksBySandbox.get(sandboxId)
+            );
+            sandboxViews.put(sandboxId, sandboxLevelResourceUsageView);
+        }
+        return sandboxViews;
+    }
+
+    /**
+     * Groups tasks by their associated sandbox.
+     *
+     * @return Map of tasks grouped by sandbox
+     */
+    private Map<String, List<Task>> getTasksGroupedBySandbox() {
+        return taskResourceTrackingService.getResourceAwareTasks()
+            .values()
+            .stream()
+            .filter(SandboxTask.class::isInstance)
+            .map(SandboxTask.class::cast)
+            .collect(Collectors.groupingBy(SandboxTask::getSandboxId, Collectors.mapping(task -> (Task) task, Collectors.toList())));
+    }
+
+    /**
+     * Calculates the resource usage of each sandbox.
+     *
+     * @param tasksBySandbox Map of tasks grouped by sandbox
+     * @return Map of resource usage for each sandbox
+     */
+    private Map<String, Map<SandboxResourceType, Long>> getResourceUsageOfSandboxes(Map<String, List<Task>> tasksBySandbox) {
+        Map<String, Map<SandboxResourceType, Long>> resourceUsageOfSandboxes = new HashMap<>();
+
+        // Iterate over each sandbox entry
+        for (Map.Entry<String, List<Task>> sandboxEntry : tasksBySandbox.entrySet()) {
+            String sandboxId = sandboxEntry.getKey();
+            List<Task> tasks = sandboxEntry.getValue();
+
+            // Prepare a usage map for the current sandbox, or retrieve the existing one
+            Map<SandboxResourceType, Long> sandboxUsage = resourceUsageOfSandboxes.computeIfAbsent(sandboxId, k -> new HashMap<>());
+
+            // Accumulate resource usage for each task in the sandbox
+            for (Task task : tasks) {
```

Comment:
I only see a single put in my code 
```java
                for (SandboxResourceType resourceType : TRACKED_RESOURCES) {
                    long currentUsage = sandboxUsage.getOrDefault(resourceType, 0L);
                    long taskUsage = resourceType.getResourceUsage(task);
                    sandboxUsage.put(resourceType, currentUsage + taskUsage);
                }
```
I prefer mine since its more readable however

Reply:
This put is happening for all the tasks in the query group while the suggested one has **only one put and no get** operation.